Add framework for multi-node requests

gavin-norman-sociomantic opened the pull request.

gavin-norman-sociomantic commented:
Super rough, initial implementation that compiles.

gavin-norman-sociomantic commented:
Based on #199.

gavin-norman-sociomantic commented:
Updated: tidied up. No longer WIP. Still blocked, as it's based on #199.

gavin-norman-sociomantic commented:
Dependency merged. Unblocked.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/RequestCore.d:
```diff
@@ -20,8 +20,15 @@ module swarm.neo.client.mixins.RequestCore;
 public enum RequestType
 {
+    // A request that operates on one node at a time. The node currently being
+    // operated on may change over the request's lifetime.
```
DDOC comments should be used instead.

gavin-norman-sociomantic replied:
Oh yeah, I missed that.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/RequestCore.d:
```diff
@@ -20,8 +20,15 @@ module swarm.neo.client.mixins.RequestCore;
     SingleNode,
+
+    // A request that operates on one node at a time. If the request fails on
+    // one node, it is repeated on the next in a round-robin sequence.
```
Does the request need to fail to be repeated on the next in a round-robin sequence? Is there are request that always tries on the next node in a round-robin sequence (like `Produce`)?

gavin-norman-sociomantic replied:
I think currently not, but I can make the comment more generic.

gavin-norman-sociomantic commented:
Updated with suggestions re: comments.

gavin-norman-sociomantic commented:
I just noticed there's loads of TODOs in here. Will fix and update.

gavin-norman-sociomantic commented:
Updated: fixed doc TODOs.

nemanja-boric-sociomantic commented:
Let's merge #211 first and rebase this, so we can confirm that the tests are passing.

nemanja-boric-sociomantic commented:
It would be helpful in this case to run `git rebase -i -x 'git commit --amend --date=$(date +%s) --no-edit' upstream/v4.x.x` to get the commit order right in GitHub as well.

Commit:
Expand handler function alias documentation

Commit:
Document different client request types

Commit:
Remove unused class field

Commit:
Reset connection counters before usage

It's cleaner to only rely on counts for the current test, not also taking
into account counts from previous tests.

Commit:
Add support for client multi-node requests

Fixes #135.

Commit:
Add missing method doc

Commit:
Correct out-dated method doc

gavin-norman-sociomantic commented:
Rebased on v4.x.x.

nemanja-boric-sociomantic commented:
That's better, the tests are red ;-)

gavin-norman-sociomantic commented:
Fixed D2 error.

nemanja-boric-sociomantic commented on src/swarm/neo/client/RequestSet.d:
```diff
@@ -263,6 +265,8 @@ public final class RequestSet: IRequestSet
             suspended fiber, waiting for I/O to complete, or may have finished.
 
             Params:
+                RequestContext = type of request context. Must be a type that
```
I'll not insist to put this in a separate commit.

gavin-norman-sociomantic replied:
Surely it counts as "Add missing method doc"?

nemanja-boric-sociomantic commented on integrationtest/neo/client/Client.d:
```diff
@@ -90,6 +93,32 @@ public class Client
+        /***********************************************************************
+
+            Assigns a DoublePut request, instructing the node to associate the
+            specified value and key.
```
Perhaps document that this puts the `(key, value)` pair into two nodes.

nemanja-boric-sociomantic commented on integrationtest/neo/client/Client.d:
```diff
@@ -276,6 +305,63 @@ public class Client
+
+            Assigns a DoublePut request, instructing two nodes to associate the
+            specified value and key. The calling Task is blocked until the
+            request finishes.
```
(re: previous comment) - just copy the docs from here.

gavin-norman-sociomantic replied:
Mm do you mean copy "Assigns a DoublePut request, instructing two nodes to associate the specified value and key." to line 98? The doc is identical.

nemanja-boric-sociomantic replied:
Yes, to copy that, but no, the doc is not identical as far as I can see (the doc at line 98 speaks about singular node).

gavin-norman-sociomantic replied:
Ahh sorry, yes, you're right.

gavin-norman-sociomantic replied:
Updated.

nemanja-boric-sociomantic commented on integrationtest/neo/client/request/internal/DoublePut.d:
```diff
@@ -0,0 +1,246 @@
+
+        /// The way in which the request ended. Used by the finished notifier to
+        /// decide what kind of notification (if any) to send to the user.
+// TODO
```
TODO

nemanja-boric-sociomantic commented on integrationtest/neo/client/request/internal/DoublePut.d:
```diff
@@ -0,0 +1,246 @@
+        // If the selected node is the first, start the request on the second.
+        if ( node_idx == 0 )
+            start_request_on_new_conn();
+// TODO: maybe it would be easier to specify the connection at this stage
```
more TODOs

gavin-norman-sociomantic replied:
Updated.

Commit:
Add multi-node request to test suite

A request that puts the same record to two arbitrary nodes is kind of dumb
and unrealistic, but it's a simple way to test the multi-node request
support.
